libc/sock/socketpair-nt.c

```diff
@@ -0,0 +1,136 @@
+/*-*- mode:c;indent-tabs-mode:nil;c-basic-offset:2;tab-width:8;coding:utf-8 -*-│
+│vi: set net ft=c ts=2 sts=2 sw=2 fenc=utf-8                                :vi│
+╞══════════════════════════════════════════════════════════════════════════════╡
+│ Copyright 2020 Justine Alexandra Roberts Tunney                              │
+│                                                                              │
+│ Permission to use, copy, modify, and/or distribute this software for         │
+│ any purpose with or without fee is hereby granted, provided that the         │
+│ above copyright notice and this permission notice appear in all copies.      │
+│                                                                              │
+│ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL                │
+│ WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED                │
+│ WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE             │
+│ AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL         │
+│ DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR        │
+│ PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER               │
+│ TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR             │
+│ PERFORMANCE OF THIS SOFTWARE.                                                │
+╚─────────────────────────────────────────────────────────────────────────────*/
+#include "libc/calls/internal.h"
+#include "libc/errno.h"
+#include "libc/mem/mem.h"
+#include "libc/nt/winsock.h"
+#include "libc/sock/internal.h"
+#include "libc/sock/yoink.inc"
+#include "libc/sock/sock.h"
+#include "libc/sysv/consts/fio.h"
+#include "libc/sysv/consts/o.h"
+#include "libc/sysv/consts/f.h"
+#include "libc/sysv/consts/sock.h"
+#include "libc/sysv/consts/af.h"
+#include "libc/sysv/consts/ipproto.h"
+
+textwindows int sys_socketpair_nt(int family, int type, int protocol, int sv[2]) {
+    struct sockaddr_storage ss;
+    struct sockaddr_in *sa = (struct sockaddr_in *)&ss;
+    uint32_t ss_len;
+    int spR = -1, spW = -1;
+    int opt;
+    int rc = -1;
+
+    int listensock = -1;
+    sv[0] = -1;     /* INVALID_SOCKET */
+    sv[1] = -1;     /* INVALID_SOCKET */
+
+    memset(&ss, 0, sizeof(ss));
+    if (family == AF_INET){
+        sa->sin_family = family;
+        sa->sin_addr.s_addr = htonl(0x7f000001);
+        sa->sin_port = 0;
+        ss_len = sizeof(struct sockaddr_in);
+    } else {
+        return -1;
+    }
+
+    listensock = sys_socket_nt(family, SOCK_STREAM, IPPROTO_TCP);
```

Comment:
Is it intentional to not pass through the `type` and `protocol` that was supplied? (Also on line 78.) I would imagine that if for some reason Windows can't support other types or protocols, that should be handled in the `sys_socket_nt` implementation.

Reply:
Hmm well good point. According to the Winsock doc, `WSASocket` (called from `sys_socket_nt`) handles other types/protocols. So I suspect if `socketpair()` is invoked with `type=SOCK_DGRAM` (apparently perfectly valid) the implementation is going to fail miserably when it calls `listen()` because DGRAM is a disconnected socket.
Now, it seems that `socketpair()` also support `type=SOCK_SEQPACKET`, and that should be ok since SEQPACKET is a connected socket... 
All right Let me run some tests and then I'll have to re-implement it. I'm thinking of something along those lines (but I need to make sure it works with DGRAM and SEQPACKET first):
```
  if (type==SOCK_DGRAM) {
     // create a pair of disconnected socket bound together
     fd1 = socket(AF_INET, SOCK_DGRAM, proto);
     bind(fd1, <localhost, any port>);
     fd2 = socket(AF_INET, SOCK_DGRAM, proto);
     connect(fd2, <localhost, port_where_fd1_is_bound>);
  } else if ((type == SOCK_STREAM) || (type == SOCK_SEQPACKET)) {
     // Similar to current implementation
  } else {
     // Unsupported type
  }
```